jsd/jsd.1
.\" $Id: jsd.1,v 1.6 2005/12/13 05:01:55 garbled Exp $
.\" Copyright (c) 1998, 1999, 2000
.\"	Tim Rightnour.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgment:
.\"	This product includes software developed by Tim Rightnour.
.\" 4. The name of Tim Rightnour may not be used to endorse or promote 
.\"    products derived from this software without specific prior written 
.\"    permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY TIM RIGHTNOUR ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL TIM RIGHTNOUR BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" The following requests are required for all man pages.
.Dd February 19, 2000
.Dt JSD 1
.Sh NAME
.Nm jsd
.Nd simple command scheduling daemon for remote execution
.Sh SYNOPSIS
.Nm
.Op Fl iqv
.Op Fl f Ar fanout
.Op Fl g Ar rungroup1,...,rungroupN
.Op Fl l Ar username
.Op Fl p Ar port
.Op Fl o Ar port
.Op Fl w Ar node1,...,nodeN
.Op Fl x Ar node1,...,nodeN
.Op Ar command ...
.Sh DESCRIPTION
The
.Nm
daemon is used to schedule remote execution of jobs across a
heterogenous cluster of machines.  Upon starting the daemon, it will
attempt to run a benchmark on all machines under it's control.
Once the daemon is active it listens for connections from a
.Em jsh
process, and gives each connecting process the name of a remote host
to execute a command on.  Faster hosts are preferred by the
.Nm
daemon, and the fastest of all remaining hosts will allways be given
out first.  Once a remote
.Em jsh
process has completed it's use of the remote host, it will again
contact the
.Nm
daemon, and return the node to service.  The daemon will never give
the same node out to more than one process at a time.
.Pp
.Nm
logs errors and critical conditions to the syslog with the facility
.Dv LOG_DAEMON .
.Pp
The following options are available:
.Bl -tag -width www
.It Fl i
The
.Fl i
option will list information about the current cluster, and command groupings. 
It will print out the current value of the fanout, and how many groups of 
machines there are within the cluster. It will also show you which command 
you are about to run, and your username if specified with the
.Fl l
option.
.It Fl q
The
.Fl q
option does not issue any commands, but displays information about the 
cluster, and the fanout groupings.
.It Fl v
Prints the version of ClusterIt to the stdout, and exits.
.It Fl f
If the
.Fl f
option is specified, followed by a number, it sets the fanout size of the 
cluster.  The fanout size is the number of nodes a command will run on in 
parallel at one time.  Thus a 80 node cluster, with a fanout size of 64, 
would run 64 nodes in parallel, then, when all have finished, it would 
execute the command on the last 16 nodes.  The fanout size defaults to 64. 
This option overrides the
.Ev FANOUT
environment variable.
.It Fl g
If the
.Fl g
option is specified, followed by a comma separated list of group names, the 
command will only be run on that group of nodes.  A node may be a part of 
more than one group if desired, however running without the
.Fl g
option will run the command on the same node as many times as it appears in the
file specified by the
.Ev CLUSTER
environment variable.  This option is silently ignored if used with the
.Fl w
option.
.It Fl l
If the
.Fl l
option is specified, followed by a username, the commands will be run under 
that userid on the remote machines.  Consideration must be taken for proper 
authentication, for this to work.
.It Fl p
The
.Fl p
option can be used to set the port number which
.Nm
will listen to
replies from
.Em jsh
processes indicating completion of processing on remote nodes.  This
option overrides the
.Ev JSD_IPORT
environment variable.  The default listen port is 2001.
.It Fl o
The
.Fl o
option can be used to set the port number which
.Nm
will listen to
requests from
.Em jsh
processes for remote nodes.  This option overrides the
.Ev JSD_OPORT
environment variable.  The default listen port is 2002.
.It Fl w
If the
.Fl w
option is specified, followed by a comma delimited list of machine names,
the
.Ar command
will be run on each node in the list.  Without this option,
.Nm
runs on the nodes listed in the file pointed to by the
.Ev CLUSTER
environment variable.
.It Fl x
The
.Fl x
option can be used to exclude specific nodes from the cluster.  The format 
is the same as the
.Fl w
option, a comma delimited list of machine names.  This option is silently 
ignored if used with the
.Fl w
option.
.El
.Sh ENVIRONMENT
.Nm
utilizes the following environment variables:
.Bl -tag -width "JSD_BENCH_CMD"
.It Ev CLUSTER
Contains a filename, which is a newline separated list of nodes
in the cluster.
.It Ev RCMD_CMD
Command to use to connect to remote machines.  The command chosen must
be able to connect with no password to the remote host.  Defaults to
.Ic rsh .
This is only used by
.Nm
when issuing the benchmark at startup, and has no
effect on how
.Em jsh
processes will connect to the remote nodes.
.It Ev RCMD_CMD_ARGS
Arguments to pass to the remote shell command.  Defaults to none. As with
.Ev RCMD_CMD
this is only used when issuing the benchmark at startup.
.It Ev JSD_BENCH_CMD
This is the name of a command, which when run on the remote nodes of
the cluster will give some indication of the relative speed of the
machines.  It does not need to be precise.  If no benchmark command is 
specified, all machines are considered to be the same speed, and will
not be wieghted.  The bench command expects a single floating point
number as the only output to the remote command.  Larger numbers are
to indicate faster machines.  See the
.Em EXAMPLES
section for more details.
.It Ev FANOUT
When set, limits the maximum number of concurrent commands sent at once.  
This can be used to keep from overloading a small host when sending out 
commands in parallel.  Defaults to 64.  This environment setting can be 
overridden by the
.Fl f
option.
.It Ev JSD_OPORT
When set, overrides the default port which
.Nm
listens on for
.Em jsh
processes requesting nodes.  This environment setting can be overriden 
by the
.Fl o
option.
.It Ev JSD_IPORT
When set, overrides the default port which
.Nm
listens on for
.Em jsh
processes reporting back that it has completed processing on a node.
This environment setting can be overriden
by the
.Fl p
option.
.El
.Sh EXAMPLES
In order to generate a benchmark figure useful to
.Nm
you will most likely need to process the ouput of a benchmark command
in some way.  There are two simple ways of doing this:
.Pp
The first way is to build a script, which exists on each machine, and
run that as your benchmark command.  The below script shows how to
process the output of the
.Xr whetstone 1
command into a format useable by
.Nm
:
.Bd -literal -offset indent
#!/bin/sh
 ./whetstone 100000 | grep MIPS | sed -e 's|.*: \\(.*\\) MIPS|\\1|'
.Ed
.Pp
The other way, would be to run a more involved benchmark on each
machine, (or assign numbers arbitrarily) and place them in a file on
the remote machines.  A benchmark command could then simply be
.Ic cat /etc/benchmark
or some other similar setup.
.Sh SEE ALSO
.Xr dsh 1 ,
.Xr jsh 1 ,
.Xr syslogd 8 .
.Sh HISTORY
The
.Nm
command first appeared in clusterit 2.0.  It should be pronounced
.Qq jez-dee ,
and stands for
.So
Job Scheduling Daemon
.Sc .
.Sh AUTHOR
.Nm Jsd
was written by Tim Rightnour.


CHANGES
$Id: CHANGES,v 1.15 2005/12/13 08:04:25 garbled Exp $

Add a -v option to all programs to tell the user what version of
ClusterIt they are running.  CVS will allways have _BETA tagged on the end,
and it will be removed at release.

Add RVT_CMD_ARGS to dvt, to allow passing of args to rvt

Incorporate patches from debian xvt package into rvt:    
Fix xvt manpage typos
Add -rv (reverse-video) option 
Better colormap handling
All other patches were either irrelevant (makefile fixes, etc) or allready
incorporated.

Fix pdf to work on linux hosts

Replace most instances of sprintf with snprintf. Did not touch rvt, as I would
prefer to keep that as unmodified as possible, and most of them would not
be easy to fix anyhow.

Replace most instances of malloc with calloc.   Mostly to prevent any
unanticipated wierdness, and a few of the mallocs were immediately followed
by memset commands anyhow.

Add RCMD_CMD_ARGS to most of the programs. This makes it easier to do
things like run "ssh -4" as your RCMD_CMD.

Stop allocating the command to be run on the remote end as buf[MAXBUF],
instead compute it at runtime and dynamically allocate it to prevent
possible overflow on a huge command.

Fix a bug noted by Jot Powers, where trailing whitespace on a GROUP or LUMP
entry would cause dsh -g to not match it.

Fix a bug noted by John Bales  where linux machines often recieved truncated 
output from the child ssh/rsh process.

Fix bug where all the programs would mangle argv[0] for ps.

* Release 2.3.1 *

Fix a bug noted by James F Wilkus on OpenBSD where dsh would segv when
reading the clusterfile.

Convert seq over to rseq fully, not just for linux anymore.  Sigh.

Fix the clusterit.spec file to work properly.  Tested on redhat

* Release 2.3 *

Made all remote commands use poll instead of a while loop to read the incoming
stream of data from stdout and stderr.  This fixes bug where linux ssh will
hang up.

Fixed linux command line argument processing.

Changed the -l <username> processing to invoke RCMD_CMD with user@host rather
than -l user host.  This makes it much more reliable.

Added an option to dsh to test remote ports before connecting.  This can be
useful where a remote host might be down, and a long timeout would occur
waiting for a response.

Updated rvt to use a newer set of sources from xvt.  Fixes questionable
licensing in xvt, and makes it more reliable.

Converted makefiles to autoconf for ease of porting.  No more stupid patch
sets which were a terrible burden to maintain and generate.

ANSIfyed most of the code (other than rvt). Nobody uses KNF anymore.. sigh..

Incorporated many patches sent by users from the internet.  I have
unfortunately lost track of what patch came from who, but thier help is
greatly appreciated.  Most of these patches either fixed bugs, added stability
or fixed porting bugs.  The xvt conversion was based mainly on a patch sent
in by a user.  My thanks go out to those of you who have sent them in.

# changes since 2.2

* Release 2.2 *
Missing this data, sorry
* Release 2.0 *

Added dvt and rvt..  an interactive form of dsh with xterms.

Fixed a bug in jsd and dsh where it would lock up running ssh waiting for
input.

Add a jsd dir, and move jsd.c from dsh to the jsd dir.  Add nearly complete
jsh program, and jcommon.[ch].

Add new common subdir and move a tonnage of crap in there.  Mainly
common.c, common.h and strsep.c.

Fix horrid bug in dsh where dsh | more would lock terminals up.  Still
need to fix in seq, run, probably pcp.

Add <netinet/if_ether.h> to barrierd.c to shut AIX C up.

Lots of minor bugfixes, mostly dealing with failures to open files, opening
0 byte clusters, etc etc..

Pcp was still using it's own code to parse strings and groups.  Fully rewrote
the back-end internals to use the new linked-list stuff, and utilize common.c.
Pcp's code is now virtually identical to dsh, whereas before it had diverged
significantly.

Error handling has finally graduated to where I no longer feel that it
"needs work".

Fix a place where malloced data was strcatted to without being bzeroed.

Group, exclusion and -w nodes are now dynamicly generated arrays.  Removes
the MAX_GROUPS restrictions.

The nodelist and grouplists are no longer maintained in any way similar to
the original setup.  Nodes are not stored in a flat array of nodes, with a
corresponding flat array of group names.  Nodes are now stores in a linked
list of machines to run on.  Nodes which are not to be run on, are not
included in this list, so the back-end routines only need to execute code
on every node they are handed in the linked list.  This is all handled in
common.c now.  This removes the all compile time restrictions on MAXNODES,
allowing the program to scale much more efficiently.

# changes since 1.3
* Release 1.3 *

Break dsh, run and seq's common code into a common.c and common.h file,
so I don't forget to add options to run and seq when fixing dsh.  Also, fix
some locations where run and seq were out of sync slightly.

Fix a bug in pdf where if the -m option was used, filesystems at 100%
would not be displayed.

Handle ^C's correctly.  If a remote process is hung, and it is the current
process being read, you may ^C it, and the interrupt will be sent to the
remote process.  This would be just like hitting ^C on the remote end, 
and will only affect the process you are currently reading from.

Added ability to pcp multiple files and wildcards to remote hosts.

Made pcp and dsh compile on an ANSI compiler (IBM C, SUN C)

Make output from dsh look nicer, and format well.

# changes since 1.2
* Release 1.2 *

Bug fixes in dsh to deal with the start of a fangroup, and
a race condition, where large output would cause it to hang
waiting for descriptors to die.

Fix a bug in dsh that caused it to lose data on AIX, and 
dump core when transitioning fangroups.

Added the rungroup, and exclusion commands to all the programs.

Fix pdf to work on a heterogeneous cluster.


jsd/jsd.c
/* $Id: jsd.c,v 1.17 2005/12/13 05:09:06 garbled Exp $ */
/*
 * Copyright (c) 2000
 *	Tim Rightnour.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by Tim Rightnour.
 * 4. The name of Tim Rightnour may not be used to endorse or promote 
 *    products derived from this software without specific prior written 
 *    permission.
 *
 * THIS SOFTWARE IS PROVIDED BY TIM RIGHTNOUR ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL TIM RIGHTNOUR BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/* Semi-intelligent job scheduling daemon.  Intended for heterogenous
 * network load sharing applications.
 *
 */

#include <errno.h>
#include <syslog.h>
#include <signal.h>
#include <strings.h>
#include <sys/types.h>
#include <sys/wait.h>

#include "../common/common.h"
#include "../common/sockcommon.h"

#if !defined(lint) && defined(__NetBSD__)
__COPYRIGHT(
"@(#) Copyright (c) 2000\n\
        Tim Rightnour.  All rights reserved\n");
__RCSID("$Id: jsd.c,v 1.17 2005/12/13 05:09:06 garbled Exp $");
#endif

/* globals */
int debug, errorflag, exclusion, grouping, iportnum, oportnum;
int nrofrungroups;
group_t *grouplist;
node_t *nodelink;
char **rungroup;
char **lumplist;
pid_t currentchild;
char *progname;

void _log_bailout(int line, char *file);
void do_bench_command(char *argv, int fanout, char *username);
void sig_handler(int i);
void main_loop(void);
void free_node(int sock);

/*
 * Main should handle deciding which group or nodes we run on, running the
 * initial benchmark run, and finally launching the scheduler daemon.
 */

int
main(int argc, char **argv)
{
    extern char	*optarg;
    extern int	optind;
    extern char *version;

    int someflag, ch, i, fanout, showflag, fanflag;
    char *p, *q, *group, *nodename, *username, *benchcmd;
    char **grouptemp, **exclude;
    struct rlimit limit;
    pid_t pid;

    someflag = showflag = fanflag = 0;
    exclusion = debug = 0;
    iportnum = oportnum = nrofrungroups = 0;
    fanout = DEFAULT_FANOUT;
    nodename = NULL;
    username = NULL;
    group = NULL;
    nodelink = NULL;

    rungroup = calloc(GROUP_MALLOC, sizeof(char **));
    if (rungroup == NULL)
	bailout();
    exclude = calloc(GROUP_MALLOC, sizeof(char **));
    if (exclude == NULL)
	bailout();

    progname = p = q = strdup(argv[0]);
    while (progname != NULL) {
	q = progname;
	progname = (char *)strsep(&p, "/");
    }
    progname = q;

#if defined(__linux__)
    while ((ch = getopt(argc, argv, "+?diqf:g:l:vw:x:p:")) != -1)
#else
    while ((ch = getopt(argc, argv, "?diqf:g:l:vw:x:p:")) != -1)
#endif
	switch (ch) {
	case 'd':		/* we want to debug jsd (hidden)*/
	    debug = 1;
	    break;
	case 'i':		/* we want tons of extra info */
	    debug = 1;
	    break;
	case 'l':		/* invoke me as some other user */
	    username = strdup(optarg);
	    break;
	case 'q':		/* just show me some info and quit */
	    showflag = 1;
	    break;
	case 'f':		/* set the fanout size */
	    fanout = atoi(optarg);
	    fanflag = 1;
	    break;
	case 'g':		/* pick a group to run on */
	    i = 0;
	    grouping = 1;
	    for (p = optarg; p != NULL; ) {
		group = (char *)strsep(&p, ",");
		if (group != NULL) {
		    if (((i+1) % GROUP_MALLOC) != 0) {
			rungroup[i++] = strdup(group);
		    } else {
			grouptemp = realloc(rungroup,
					    i*sizeof(char **) +
					    GROUP_MALLOC*sizeof(char *));
			if (grouptemp != NULL)
			    rungroup = grouptemp;
			else
			    bailout();
			rungroup[i++] = strdup(group);
		    }
		}
	    }
	    nrofrungroups = i;
	    group = NULL;
	    break;			
	case 'x':		/* exclude nodes, w overrides this */
	    exclusion = 1;
	    i = 0;
	    for (p = optarg; p != NULL; ) {
		nodename = (char *)strsep(&p, ",");
		if (nodename != NULL) {
		    if (((i+1) % GROUP_MALLOC) != 0) {
			exclude[i++] = strdup(nodename);
		    } else {
			grouptemp = realloc(exclude,
					    i*sizeof(char **) +
					    GROUP_MALLOC*sizeof(char *));
			if (grouptemp != NULL)
			    exclude = grouptemp;
			else
			    bailout();
			exclude[i++] = strdup(nodename);
		    }
		}
	    }
	    break;
	case 'w':		/* perform operation on these nodes */
	    someflag = 1;
	    for (p = optarg; p != NULL; ) {
		nodename = (char *)strsep(&p, ",");
		if (nodename != NULL)
		    (void)nodealloc(nodename);
	    }
	    break;
	case 'o':		/* port to listen to requests on */
	    oportnum = atoi(optarg);
	    break;
	case 'p':		/* port to listen to completions on */
	    iportnum = atoi(optarg);
	    break;
	case 'v':
	    (void)printf("%s: %s\n", progname, version);
	    exit(EXIT_SUCCESS);
	    break;
	case '?':		/* you blew it */
	    (void)fprintf(stderr,
		 "usage: jsd [-iqv] [-f fanout] [-g rungroup1,...,rungroupN] "
		 "[-l username] [-p port] [-o port] [-w node1,..,nodeN] "
       		 "[-x node1,...,nodeN] [command ...]\n");
	    return(EXIT_FAILURE);
	    /*NOTREACHED*/
	    break;
	default:
	    break;
	}

/* check for a fanout var, and use it if the fanout isn't on the commandline */
    if (!fanflag)
	if (getenv("FANOUT"))
	    fanout = atoi(getenv("FANOUT"));
    if (!iportnum) {
	if (getenv("JSD_IPORT"))
	    iportnum = atoi(getenv("JSD_IPORT"));
	else
	    iportnum = JSDIPORT;
    }
    if (username == NULL)
        if (getenv("RCMD_USER"))
            username = strdup(getenv("RCMD_USER"));
    if (!oportnum) {
	if (getenv("JSD_OPORT"))
	    oportnum = atoi(getenv("JSD_OPORT"));
	else
	    oportnum = JSDOPORT;
    }
    if (!someflag)
	parse_cluster(exclude);

    argc -= optind;
    argv += optind;
    if (showflag) {
	do_showcluster(fanout);
	return(EXIT_SUCCESS);
    }
    openlog("jsd", LOG_PID|LOG_NDELAY, LOG_DAEMON);

    /*
     * set per-process limits for max descriptors, this avoids running
     * out of descriptors and the odd errors that come with that.
     */
    close(STDIN_FILENO);
    pid = fork();
    if (pid == 0) {
	(void)setsid(); /* disassociate from our controlling terminal */
	syslog(LOG_INFO, "Job Scheduling Daemon Started");

	if (getrlimit(RLIMIT_NOFILE, &limit) != 0)
	    bailout();
	if (limit.rlim_cur < fanout * 5) {
	    limit.rlim_cur = fanout * 5;
	    if (setrlimit(RLIMIT_NOFILE, &limit) != 0)
		bailout();
	}
		
	if (getenv("JSD_BENCH_CMD")) {
	    benchcmd = getenv("JSD_BENCH_CMD");
	    do_bench_command(benchcmd, fanout, username);
	} else {
	    syslog(LOG_WARNING, "No JSD_BENCH_CMD environment setting,"
		" assuming homogenus cluster.");
	}

	/* jump to the loop */
	main_loop();
    } else if (pid > 0) {
	(void)printf("%d\n", pid); /* spit the pid out */
	return(EXIT_SUCCESS);
    }
    else if (pid == -1)
	syslog(LOG_CRIT, "Aborting: %m");
    return(EXIT_FAILURE);
}

/*
 * This performs the bulk of the program's purpose.  The program enters this
 * loop, and either watches the shared memory, and talks with jsh processes,
 * or listens on the jsd port, and listens for jsh processes asking for a
 * node handout.
 *
 * The theory of operation is simple:
 * a jsh process is invoked to run a command, similar to run.  jsh contacts
 * the jsd asking for an available node.  jsd locks the database, assigns a
 * node, marks that node as in use, and unlocks the db.  Eventually the
 * jsh process will complete it's task, and report back to jsd.  Jsd will
 * then free up the node for further processing.
 */

void main_loop(void)
{
    char *buf;
    node_t *nodeptr, *fastnode;
    double topspeed;
    int osock, isock, new;
    struct sockaddr_in clientname;
    size_t size;
    fd_set node_fd_set, free_fd_set, full_fd_set;
    struct timeval timeout;
    struct sigaction signaler;

    buf = NULL;

    signaler.sa_handler = sig_handler;
    signaler.sa_flags = SA_RESTART;
    sigemptyset(&signaler.sa_mask);
    if (sigaction(SIGTERM, &signaler, NULL) != 0)
	log_bailout();
	
    osock = make_socket(oportnum);
    isock = make_socket(iportnum);
#ifdef HAVE_SETPRIORITY
       /* not implemented in Cygwin */
       /* read: http://www.cygwin.com/ml/cygwin/2003-10/msg00664.html */
    setpriority(PRIO_PROCESS, 0, 20); /* nice ourselves */
#endif /* HAVE_SETPRIORITY */

    if (listen(osock, SOMAXCONN) < 0)
	log_bailout();
    if (listen(isock, SOMAXCONN) < 0)
	log_bailout();


    for (;;) { /* loop */
	FD_ZERO(&node_fd_set);
	FD_ZERO(&full_fd_set);
	FD_ZERO(&free_fd_set);
	FD_SET(osock, &node_fd_set);
	FD_SET(isock, &free_fd_set);
	FD_SET(osock, &full_fd_set);
	FD_SET(isock, &full_fd_set);
	timeout.tv_sec = 0;
	timeout.tv_usec = 0;
	if (debug)
	    syslog(LOG_DEBUG, "Entering main loop");
	if (select(FD_SETSIZE, &full_fd_set, &full_fd_set, NULL, NULL) < 0)
	    log_bailout();
	if (debug)
	    syslog(LOG_DEBUG, "We have a connection");
	if (select(FD_SETSIZE, &free_fd_set, &free_fd_set, NULL, &timeout)) {
	    /* jsh wants to free a node */
	    if (debug)
		syslog(LOG_DEBUG, "Someone wants to free a node");
	    free_node(isock);
	} else if (select(FD_SETSIZE, &node_fd_set, NULL, NULL, &timeout)) {
	    /* jsh wants a node */
	    if (debug)
		syslog(LOG_DEBUG, "Someone wants a node");
	    new = accept(osock, (struct sockaddr *) &clientname, &size);
	    if (new < 0)
		log_bailout();
	    topspeed = 0.0;
	    fastnode = nodeptr = NULL;
	    while (fastnode == NULL)
		for (nodeptr = nodelink; nodeptr; nodeptr = nodeptr->next) {
		    FD_ZERO(&free_fd_set);
		    FD_SET(isock, &free_fd_set);
		    if (nodeptr->index > topspeed && nodeptr->free)
			fastnode = nodeptr;
		    else if (select(FD_SETSIZE, &free_fd_set, &free_fd_set,
				 NULL, &timeout)) {
			free_node(isock);
		    }
		}
	    fastnode->free = 0;
	    if (debug)
		syslog(LOG_DEBUG, "Handing out node %s to a jsh process",
		    fastnode->name);
	    write_to_client(new, fastnode->name);
	    close(new);
	}
    }
}

/*
* Free up a node for use by other requestors
*/

void free_node(int sock)
{
    int new, i;
    struct sockaddr_in clientname;
    size_t size;
    char *buf;
    node_t *nodeptr;

    if (debug)
	syslog(LOG_DEBUG, "Entered free_node");
    new = accept(sock, (struct sockaddr *) &clientname, &size);
    if (debug)
	syslog(LOG_DEBUG, "accepted new connection");
    if (new < 0)
	log_bailout();
    write_to_client(new, "1");
    i = read_from_client(new, &buf);
    buf[i] = '\0';
    if (debug)
	syslog(LOG_DEBUG, "got node %s from client", buf);
    for (nodeptr = nodelink; nodeptr; nodeptr = nodeptr->next)
	if (strcmp(buf, nodeptr->name) == 0) {
	    nodeptr->free = 1;
	    if (debug)
		syslog(LOG_DEBUG, "freeing node %s", nodeptr->name);
	}
    close(new);
}

/*
 * Note that while the below is nearly identical, it has but one purpose in
 * life, and that is to populate the index of machine speeds.
 */

void
do_bench_command(char *argv, int fanout, char *username)
{
    FILE *fd, *in;
    char pipebuf[2048], buf[MAXBUF];
    int status, i, j, n, g, nrofargs, arg;
    char *q, *rsh, *cd, *p, *rshargs, **cmd;
    node_t *nodeptr, *nodehold;

    j = i = 0;
    in = NULL;
    q = NULL;
    cd = pipebuf;

    if (debug) {
	if (username != NULL)
	    syslog(LOG_DEBUG, "As User: %s", username);
	syslog(LOG_DEBUG, "On nodes:");
    }
    for (nodeptr = nodelink; nodeptr; nodeptr = nodeptr->next) {
	if (debug) {
	    q = (char *)calloc(MAXBUF, sizeof(char));
	    if (q == NULL)
		log_bailout();
	    if (!(j % 4) && j > 0)
		strcat(q, "\n");
	    strcat(q, nodeptr->name);
	    strcat(q, "\t");
	}
	j++;
    }
    if (debug)
	syslog(LOG_DEBUG, "%s", q);

    i = j; /* side effect of above */
    j = i / fanout;
    if (i % fanout)
	j++; /* compute the # of rungroups */

    if (debug) {
	syslog(LOG_DEBUG, "\nDo Command: %s", argv);
	syslog(LOG_DEBUG, "Fanout: %d Groups:%d", fanout, j);
    }

    rsh = getenv("RCMD_CMD");
    if (rsh == NULL)
	rsh = strdup("rsh");
    if (rsh == NULL)
	bailout();
    if (getenv("RCMD_CMD_ARGS") != NULL)
	rshargs = strdup(getenv("RCMD_CMD_ARGS"));
    nrofargs = 3;
    if (rshargs != NULL) {
	p = rshargs;
	nrofargs++;
	while (*p != '\0') {
	    if (isspace(*p))
		nrofargs++;
	    *p++;
	}
    }

    g = 0;
    nodeptr = nodelink;
    for (n=0; n <= j; n++) {
	nodehold = nodeptr;
	for (i=0; (i < fanout && nodeptr != NULL); i++) {
	    g++;
	    if (debug)
		syslog(LOG_DEBUG, "Working node: %d, fangroup %d,"
		    " fanout part: %d", g, n, i);
/*
 * we set up pipes for each node, to prepare for the oncoming barrage of data.
 * Close on exec must be set here, otherwise children spawned after other
 * children, inherit the open file descriptors, and cause the pipes to remain
 * open forever.
 */
	    if (pipe(nodeptr->out.fds) != 0)
		log_bailout();
	    if (fcntl(nodeptr->out.fds[0], F_SETFD, 1) == -1)
		log_bailout();
	    if (fcntl(nodeptr->out.fds[1], F_SETFD, 1) == -1)
		log_bailout();
	    nodeptr->childpid = fork();
	    switch (nodeptr->childpid) {
		/* its the ol fork and switch routine eh? */
	    case -1:
		log_bailout();
		break;
	    case 0:
		/* remove from parent group to avoid kernel
		 * passing signals to children.
		 */
		(void)setsid();
		if (dup2(nodeptr->out.fds[1], STDOUT_FILENO) 
		    != STDOUT_FILENO) 
		    log_bailout();
		if (close(nodeptr->out.fds[0]) != 0)
		    log_bailout();

		if (username != NULL)
		    (void)snprintf(buf, MAXBUF, "%s@%s", username,
				   nodeptr->name);
		else
		    (void)snprintf(buf, MAXBUF, "%s", nodeptr->name);
		if (debug)
		    (void)syslog(LOG_DEBUG, "%s %s %s %s\n", rsh,
			     (rshargs? rshargs:""), buf, argv);
		cmd = calloc(nrofargs+1, sizeof(char *));
		arg = 0;
		cmd[arg++] = rsh;
		while (rshargs != NULL)
		    cmd[arg++] = strdup(strsep(&rshargs, " "));
		cmd[arg++] = buf;
		cmd[arg++] = argv;
		cmd[arg] = (char *)0;
		execvp(rsh, cmd);
		log_bailout();
		break;
	    default:
		break;
	    } /* switch */
	    nodeptr = nodeptr->next;
	} /* for i */
	nodeptr = nodehold;
	for (i=0; (i < fanout && nodeptr != NULL); i++) {
	    if (debug)
		syslog(LOG_DEBUG, "Printing node: %d, fangroup %d,"
		    " fanout part: %d", g-fanout+i, n, i);
	    currentchild = nodeptr->childpid;
	    /* now close off the useless stuff, and read the goodies */
	    if (close(nodeptr->out.fds[1]) != 0)
		log_bailout();
	    fd = fdopen(nodeptr->out.fds[0], "r"); /* stdout */
	    if (fd == NULL)
		log_bailout();
	    while ((cd = fgets(pipebuf, sizeof(pipebuf), fd))) {
		if (cd != NULL) {
		    nodeptr->index = atof(cd);
		    syslog(LOG_DEBUG, "recorded speed %f for node %s",
			nodeptr->index, nodeptr->name);
		}
	    }
	    fclose(fd);
	    (void)wait(&status);
	    nodeptr = nodeptr->next;
	} /* for i */			
    } /* for n */
}

/*ARGSUSED*/
void
_log_bailout(int line, char *file)
{
    syslog(LOG_CRIT, "%s: Failed in %s line %d: %m %d", progname, file, line,
	errno);

    _exit(EXIT_FAILURE);
}

void
sig_handler(int i)
{

    switch (i) {
    case SIGTERM:
	_exit(EXIT_SUCCESS);
	break;
    default:
	log_bailout();
	break;
    }
}
